<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#Источник-данных" data-toc-modified-id="Источник-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Источник данных</a></span></li><li><span><a href="#Материалы" data-toc-modified-id="Материалы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Материалы</a></span></li><li><span><a href="#Подготовка-инструментов-и-загрузка-данных" data-toc-modified-id="Подготовка-инструментов-и-загрузка-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка инструментов и загрузка данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Пропущенные-значения" data-toc-modified-id="Пропущенные-значения-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Пропущенные значения</a></span></li><li><span><a href="#Типы-данных" data-toc-modified-id="Типы-данных-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Типы данных</a></span></li><li><span><a href="#Уникальные-значения" data-toc-modified-id="Уникальные-значения-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Уникальные значения</a></span></li></ul></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Расчет-времени-ожидания-ответа" data-toc-modified-id="Расчет-времени-ожидания-ответа-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Расчет времени ожидания ответа</a></span></li><li><span><a href="#Расчет-calls_rate" data-toc-modified-id="Расчет-calls_rate-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Расчет calls_rate</a></span></li><li><span><a href="#Обзор-возможных-аномалий" data-toc-modified-id="Обзор-возможных-аномалий-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Обзор возможных аномалий</a></span></li><li><span><a href="#Оценка-средних-значений-calls_rate-и--response_time-для-входящих-вызовов-на-оператора-в-день" data-toc-modified-id="Оценка-средних-значений-calls_rate-и--response_time-для-входящих-вызовов-на-оператора-в-день-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Оценка средних значений calls_rate и  response_time для входящих вызовов на оператора в день</a></span></li><li><span><a href="#Оценка-среднего-числа-исходящих-внешних-вызовов-на-оператора-в-день" data-toc-modified-id="Оценка-среднего-числа-исходящих-внешних-вызовов-на-оператора-в-день-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Оценка среднего числа исходящих внешних вызовов на оператора в день</a></span></li><li><span><a href="#Определение-коридоров-нормальных-значений-для-calls_rate,-response_time-и-числа-исходящих-вызовов-на-оператора-в-день" data-toc-modified-id="Определение-коридоров-нормальных-значений-для-calls_rate,-response_time-и-числа-исходящих-вызовов-на-оператора-в-день-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Определение коридоров нормальных значений для calls_rate, response_time и числа исходящих вызовов на оператора в день</a></span></li><li><span><a href="#Число-внутренних-исходящих-вызовов-на-оператора-в-день" data-toc-modified-id="Число-внутренних-исходящих-вызовов-на-оператора-в-день-6.7"><span class="toc-item-num">6.7&nbsp;&nbsp;</span>Число внутренних исходящих вызовов на оператора в день</a></span></li><li><span><a href="#Длительность-вызовов" data-toc-modified-id="Длительность-вызовов-6.8"><span class="toc-item-num">6.8&nbsp;&nbsp;</span>Длительность вызовов</a></span></li><li><span><a href="#Объединение-всех-показателей" data-toc-modified-id="Объединение-всех-показателей-6.9"><span class="toc-item-num">6.9&nbsp;&nbsp;</span>Объединение всех показателей</a></span></li></ul></li><li><span><a href="#Выявление-неэффективных-операторов" data-toc-modified-id="Выявление-неэффективных-операторов-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Выявление неэффективных операторов</a></span><ul class="toc-item"><li><span><a href="#Колл-центры,-где-выявлены-неэффективные-операторы" data-toc-modified-id="Колл-центры,-где-выявлены-неэффективные-операторы-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Колл-центры, где выявлены неэффективные операторы</a></span></li></ul></li><li><span><a href="#Проверка-статистических-гипотез" data-toc-modified-id="Проверка-статистических-гипотез-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Проверка статистических гипотез</a></span><ul class="toc-item"><li><span><a href="#У-неэффективных-операторов-среднее-calls_rate-выше,-чем-среднее-calls_rate-в-целом." data-toc-modified-id="У-неэффективных-операторов-среднее-calls_rate-выше,-чем-среднее-calls_rate-в-целом.-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>У неэффективных операторов среднее calls_rate выше, чем среднее calls_rate в целом.</a></span></li><li><span><a href="#У-неэффективных-операторов-среднее-response_time-выше,-чем-среднее-в-целом." data-toc-modified-id="У-неэффективных-операторов-среднее-response_time-выше,-чем-среднее-в-целом.-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>У неэффективных операторов среднее response_time выше, чем среднее в целом.</a></span></li></ul></li><li><span><a href="#Выводы-и-рекомендации" data-toc-modified-id="Выводы-и-рекомендации-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Выводы и рекомендации</a></span></li></ul></div>

# Определение неэффективных операторов колл-центров

Анализ выполняется для компании - провайдера телекоммуникационных услуг "Нупозвони".

Провайдер предоставляет колл-центрам технические и программные средства для выполнения исходящих или приема входящих вызовов, а также для связи операторов одного колл-центра между собой. Операторы являются сотрудниками колл-центров, то есть клиентов "Нупозвони".

## Цель исследования

1. Определить неэффективных операторов.
1. Предложить методику автоматического выявления операторов с низкой эффективностью.

## Источник данных

Данные собраны провайдером телекоммуникационных услуг.

## Материалы

[Презентация](https://disk.yandex.ru/i/Swyr1TmmR3qWMg "Презентация")

[Дашборд](https://public.tableau.com/views/qualification_dashboard/sheet2?:language=en-US&:display_count=n&:origin=viz_share_link "Дашборд")

## Подготовка инструментов и загрузка данных

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats as st

In [2]:
try:
    data = pd.read_csv('telecom_dataset.csv')
    clients = pd.read_csv('telecom_clients.csv')
except:
    data = pd.read_csv('/datasets/telecom_dataset.csv')
    clients = pd.read_csv('/datasets/telecom_clients.csv')
display(data.head(), clients.head(), data.info(), clients.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              53902 non-null  int64  
 1   date                 53902 non-null  object 
 2   direction            53902 non-null  object 
 3   internal             53785 non-null  object 
 4   operator_id          45730 non-null  float64
 5   is_missed_call       53902 non-null  bool   
 6   calls_count          53902 non-null  int64  
 7   call_duration        53902 non-null  int64  
 8   total_call_duration  53902 non-null  int64  
dtypes: bool(1), float64(1), int64(4), object(3)
memory usage: 3.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      732 non-null    int64 
 1   tariff_plan  732 non-null    object
 2 

user_id                       date direction internal  operator_id  \
0   166377  2019-08-04 00:00:00+03:00        in    False          NaN   
1   166377  2019-08-05 00:00:00+03:00       out     True     880022.0   
2   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
3   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
4   166377  2019-08-05 00:00:00+03:00       out    False     880022.0   

   is_missed_call  calls_count  call_duration  total_call_duration  
0            True            2              0                    4  
1            True            3              0                    5  
2            True            1              0                    1  
3           False            1             10                   18  
4            True            3              0                   25

user_id tariff_plan  date_start
0   166713           A  2019-08-15
1   166901           A  2019-08-23
2   168527           A  2019-10-29
3   167097           A  2019-09-01
4   168193           A  2019-10-16

None

None

## Предобработка данных
### Дубликаты

In [ ]:
data.duplicated().sum()

In [ ]:
data = data.drop_duplicates()

In [ ]:
data.duplicated().sum()

In [ ]:
data['operator_id'].isna().sum()

1. Полностью задублированные строки удалены (4900 штук).
1. В результате незначительно сократилось число случаев, когда неизвестен operator_id.
1. Не представляется возможным проверить датасет на наличие неявных дубликатов.

### Пропущенные значения

In [ ]:
data = data.dropna(subset=['internal', 'operator_id'])

In [ ]:
data.isna().sum()

1. Для целей более точного определения признаков неэффективности операторов предполагалось сохранить эти строки, но при проверке численности звонков на каждого оператора, мы увидели, что на фальш-значение 0 приходится более 6600 вызовов. При расчете средних значений это дало бы серьезное искажение. Строки с отсутствующими operator_id удалены.
1. Строки, в которых пропуски зафиксированы в колонке 'internal', удалены ввиду их малого количества (менее 1% всего датасета).

### Типы данных

In [ ]:
data['date'] =  pd.to_datetime(data['date'], format='%Y-%m-%dT%H:%M:%S')

In [ ]:
data['internal'] = data['internal'].astype('bool')

In [ ]:
data['operator_id'] = data['operator_id'].astype('int')

In [ ]:
data['is_missed_call'] = data['is_missed_call'].replace(True, 1)
data['is_missed_call'] = data['is_missed_call'].replace(False, 0)

In [ ]:
data.info()

In [ ]:
data.head()

1. Уточнены типы данных для колонок:
    - date, переведена в datetime;
    - internal, переведена в bool;
    - operator_id, переведена в int.
1. Для удобства расчета calls_rate отношения числа пропущенных вызовов ко всем входящим внешним - значения True / False в колонке is_missed_call заменены на 1 / 0 соответственно.
1. Попытка привести значения call_duration, total_call_duration к минутам вызывают потерю данных на строках с малыми значениями. Эти колонки оставлены как есть.

### Уникальные значения

In [ ]:
data['user_id'].nunique()

In [ ]:
data['direction'].unique()

In [ ]:
data['operator_id'].nunique()

Наборы уникальных значений выглядят адекватно.

## Исследовательский анализ данных
### Расчет времени ожидания ответа

In [ ]:
data['response_time'] = data['total_call_duration'] - data['call_duration']

Значение response_time означает время в секундах, которое прошло с момента поступления вызова до ответа оператора. Чем оно меньше, тем более эффективно работает оператор.

### Расчет calls_rate

In [ ]:
incoming_calls_by_operators = data.query('direction == "in" and internal == False')
incoming_calls_by_operators = incoming_calls_by_operators.pivot_table(index = ['date', 'operator_id', 'internal'], values = 'direction', aggfunc = 'count').reset_index()

In [ ]:
incoming_calls_response = data.query('direction == "in" and internal == False')
incoming_calls_response = incoming_calls_response.pivot_table(index = ['date', 'operator_id', 'internal'], values = 'is_missed_call', aggfunc = 'sum').reset_index()

In [ ]:
incoming_calls_rate = incoming_calls_by_operators.merge(incoming_calls_response, on = ['date', 'operator_id', 'internal'], how = 'left')


In [ ]:
incoming_calls_rate['calls_rate'] = incoming_calls_rate['is_missed_call'] / incoming_calls_rate['direction']

In [ ]:
incoming_calls_rate = incoming_calls_rate.rename(columns={'direction': 'calls_cnt'})

In [ ]:
for row in incoming_calls_rate:
    incoming_calls_rate['direction'] = 'in'
incoming_calls_rate.head()

In [ ]:
data = data.merge(incoming_calls_rate, on = ['date', 'operator_id', 'direction', 'internal'], how = 'left')

In [ ]:
data = data.drop(columns = ['calls_cnt', 'is_missed_call_y'])
data = data.rename(columns={'is_missed_call_x': 'is_missed_call'})
data.head()

Значение calls_rate показывает, насколько эффективно отвечал на входящие внешние вызовы данный оператор в данную дату. Чем меньше значение calls_rate, тем больше обработанных входящих внешних вызовов было у данного оператора в данную дату.

### Обзор возможных аномалий

In [ ]:
anomaly_overview = data.describe()
anomaly_overview = anomaly_overview.T
anomaly_overview['bottom_whisker'] = anomaly_overview['25%'] - 1.5 * (anomaly_overview['75%'] - anomaly_overview['25%'])
anomaly_overview['upper_whisker'] = anomaly_overview['75%'] + 1.5 * (anomaly_overview['75%'] - anomaly_overview['25%'])
anomaly_overview

In [ ]:
data = data.query('calls_count <= 33.5')
data = data.query('call_duration <= 1930.5')
data = data.query('total_call_duration <= 2667')
data = data.query('response_time <= 519.0')
data.info()

Удалены как выбросы значения в колонках calls_count, call_duration, total_call_duration, response_time, превышающие границу правого уса.

### Оценка средних значений calls_rate и  response_time для входящих вызовов на оператора в день

In [ ]:
incoming_values = data.query('direction == "in" and internal == False')
incoming_avg_values = incoming_values.pivot_table(index = ['operator_id', 'date'], values = ['calls_rate', 'response_time'], aggfunc = 'mean').reset_index()
incoming_avg_values.sort_values(by = 'date', ascending = True)
incoming_avg_values = incoming_avg_values.pivot_table(index = 'operator_id', values = ['calls_rate', 'response_time'], aggfunc = 'mean').reset_index()
display(incoming_avg_values)
display('Среднее значение calls_rate на оператора в день:', incoming_avg_values['calls_rate'].mean(), 'Среднее значенние response_time на оператора в день:', incoming_avg_values['response_time'].mean())

In [ ]:
plt.figure(figsize=(8, 6))

plt.hist(incoming_avg_values['calls_rate'], bins=15, color='skyblue', edgecolor='black')

plt.xlabel('Среднее значение calls_rate')
plt.ylabel('Частота')
plt.title('Распределение среднего значения calls_rate')

plt.show()

In [ ]:
plt.figure(figsize=(8, 6))

plt.hist(incoming_avg_values['response_time'], bins=15, color='skyblue', edgecolor='black')

plt.xlabel('Среднее значение response_time')
plt.ylabel('Частота')
plt.title('Распределение среднего значения response_time')

plt.show()

### Оценка среднего числа исходящих внешних вызовов на оператора в день

In [ ]:
outcoming_calls = data.query('direction == "out" and internal == False')
out_pivot = outcoming_calls.pivot_table(index = ['operator_id', 'date'], values = 'direction', aggfunc = 'count').reset_index()
out_pivot = out_pivot.pivot_table(index = 'operator_id', values = 'direction', aggfunc = 'mean').reset_index()
display('Среднее число исходящих вызовов на оператора в день:', out_pivot['direction'].mean(), out_pivot)

In [ ]:
plt.figure(figsize=(8, 6))

plt.hist(out_pivot['direction'], bins=15, color='skyblue', edgecolor='black')

plt.xlabel('Среднее число исходящих внешних вызовов')
plt.ylabel('Частота')
plt.title('Распределение среднего числа исходящих внешних вызовов')

plt.show()

In [ ]:
general_avg_values = incoming_avg_values.merge(out_pivot, on = 'operator_id', how = 'left')
general_avg_values = general_avg_values.rename(columns = {'direction' : 'out_calls_cnt'})
general_avg_values

### Определение коридоров нормальных значений для calls_rate, response_time и числа исходящих вызовов на оператора в день

In [ ]:
range_overview = general_avg_values.describe()
range_overview = range_overview.T
range_overview

1. Коридор нормальных значений для calls_rate составляет от 0.0 до 0.02.
1. Нормальный диапазон response_time: от 1 до 52 секунд.
1. Нормальное число выполненных исходящих вызовов - от 1.14 до 1.53.

### Число внутренних исходящих вызовов на оператора в день

In [ ]:
internal_calls = data.query('direction == "out" and internal == True')
int_pivot = internal_calls.pivot_table(index = ['operator_id', 'date'], values = 'direction', aggfunc = 'count').reset_index()
int_pivot = int_pivot.pivot_table(index = 'operator_id', values = 'direction', aggfunc = 'mean').reset_index()
int_pivot = int_pivot.rename(columns = {'direction' : 'int_calls_cnt'})
display('Среднее число исходящих внутренних вызовов на оператора в день:', int_pivot['int_calls_cnt'].mean(), int_pivot)

In [ ]:
plt.figure(figsize=(8, 6))

plt.hist(int_pivot['int_calls_cnt'], bins=15, color='skyblue', edgecolor='black')

plt.xlabel('Среднее число внутренних исходящих вызовов')
plt.ylabel('Частота')
plt.title('Распределение среднего числа внутренних исходящих вызовов')

plt.show()

In [ ]:
internal_calls_overview = int_pivot.describe()
internal_calls_overview = internal_calls_overview.T
internal_calls_overview

Эффективный оператор обычно делает 1 внутренний исходящий вызов в течение дня. Однако макисмальное значение для этого параметра - 2, что выглядит вполне адекватно  точки зрения здравого смысла. Поэтому мы выведем этот показатель из определения неэффективных операторов.

### Длительность вызовов

In [ ]:
# расчет средней продолжительности внешнего разговора для каждого оператора независимо от направления 
avg_call_duration_by_operator = data.query('internal == False').pivot_table(index = 'operator_id', values = 'call_duration', aggfunc = 'mean').reset_index()
avg_call_duration_by_operator

In [ ]:
plt.figure(figsize=(8, 6))

plt.hist(avg_call_duration_by_operator['call_duration'], bins=15, color='skyblue', edgecolor='black')

plt.xlabel('Средняя продолжительность разговора')
plt.ylabel('Частота')
plt.title('Распределение средней продолжительности разговора')

plt.show()

In [ ]:
# расчет общего среднего времени разговора для внешнего звонка независимо от направления
avg_call_duration = data.query('internal == False')
avg_call_duration['call_duration'].describe()

Продолжительность звонка у эффективного оператора от 1 минуты 40 секунд до почти 8,5 минут, а в среднем звонок длится 5,5 минут.

### Объединение всех показателей

In [ ]:
general_avg_values = general_avg_values.merge(int_pivot, on = 'operator_id', how = 'left')
general_avg_values = general_avg_values.merge(avg_call_duration_by_operator, on = 'operator_id', how = 'left')
general_avg_values

## Выявление неэффективных операторов

In [ ]:
underperforming = general_avg_values.query('((calls_rate > 0.02 and response_time > 55) or out_calls_cnt < 1.37) and (call_duration < 60 or call_duration > 504)')
display(underperforming)
underperforming.info()

1. В число неэффективных попадают 65 оператора по параметрам calls_rate + response_time или по числу исходящих вызовов, а также по длительности звонков.

### Колл-центры, где выявлены неэффективные операторы

In [ ]:
operator_list = list(underperforming['operator_id'])
client_list = data.query('operator_id in @operator_list and operator_id != 0')
client_list

In [ ]:
display(
    'Число колл-центров, где выявлены неэффективные операторы:',
    client_list['user_id'].nunique(),
    client_list['user_id'].unique()
)

## Проверка статистических гипотез
### У неэффективных операторов среднее calls_rate выше, чем среднее calls_rate в целом.

Напомним, что в нашем случае чем ниже calls_rate, тем более эффективен оператор.

H0: среднее значение calls_rate в целом равно calls_rate у неэффективных операторов.

H1: среднее значение calls_rate в целом ниже, чем calls_rate у неэффективных операторов.

In [ ]:
avg_calls_rate_under = round(underperforming['calls_rate'].mean(), 2)
display('Среднее значение calls_rate у неэффективных операторов:', avg_calls_rate_under)

In [ ]:
general_calls_rate = general_avg_values['calls_rate']
alpha = .05
results = st.ttest_1samp(
    general_calls_rate, 
    avg_calls_rate_under)
display('p-значение:', results.pvalue / 2)
if (results.pvalue / 2 < alpha) and (general_calls_rate.mean() < avg_calls_rate_under):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не отвергаем нулевую гипотезу")

### У неэффективных операторов среднее response_time выше, чем среднее в целом.

H0: среднее значение response_time в целом равно response_time у неэффективных операторов.

H1: среднее значение response_time в целом ниже, чем response_time у неэффективных операторов.

In [ ]:
avg_resp_time_under = round(underperforming['response_time'].mean(), 2)
display('Среднее значение response_time у неэффективных операторов:', avg_resp_time_under)

In [ ]:
general_response_time = general_avg_values['response_time']
alpha = .05
results = st.ttest_1samp(
    general_response_time, 
    avg_resp_time_under)
display('p-значение:', results.pvalue / 2)
if (results.pvalue / 2 < alpha) and (general_response_time.mean() < avg_resp_time_under):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не отвергаем нулевую гипотезу")

В обоих случаях мы убедились, что средние показатели неэффективных операторов хуже, чем средние значения показателей в целом.

## Выводы и рекомендации
1. Мы обнаружили, что в 48 колл-центрах работают операторы, чьи средние показатели эффективности отстают от средних значений в целом:
       166520, 166536, 166582, 166658, 166669, 166680, 166688, 166713,
       166774, 166782, 166813, 166874, 166901, 166916, 166939, 167009,
       167016, 167053, 167059, 167082, 167113, 167176, 167264, 167397,
       167497, 167521, 167533, 167621, 167626, 167650, 167654, 167742,
       167756, 167759, 167805, 167840, 167852, 167906, 167930, 167969,
       168008, 168062, 168187, 168225, 168252, 168307, 168481, 168500

1. Главными показателями мы выбрали отношение пропущенных вызовов к отвеченным и время ожидания ответа для входящих вызовов и число исходящих вызовов в день для операторов, работающих на исходящих вызовах.

1. Второстепенным показателем стала длительность разговора.

1. При этом мы обнаружили, что средние значения исходящих внешних вызовов в целом очень малы - не более 2 вызовов. Это позволет предположить, что наши колл-центры в целом больше заточены на прием вызовов, чем на исходящие.

1. Руководители колл-центров могут сами определить, какие показатели для них более существенны, исходя из специфики своей работы.
1. Кроме того, в зависимости от признака неэффективности разные меры могут быть применены для исправления ситуации: например, дополнительное обучение для тех, у длительность вызовов слишком высока, или более глубокий анализ причин большого числа пропущенных звонков.